pennylane/gradients/finite_difference.py

```diff
@@ -154,6 +154,211 @@ def finite_diff_coeffs(n, approx_order, strategy):
+            else:
+                dim = 1
+
+            sub_list_zeros = [qml.math.zeros(dim) for _ in range(len(tape.trainable_params))]
+            sub_list_zeros = tuple(sub_list_zeros)
+
+            list_zeros.append(sub_list_zeros)
+
+        if len(tape.measurements) == 1:
+            return [], lambda _: list_zeros[0]
+
+        return [], lambda _: tuple(list_zeros)
+
+    gradient_tapes = []
+    shapes = []
+    c0 = None
+
+    coeffs, shifts = finite_diff_coeffs(n=n, approx_order=approx_order, strategy=strategy)
+
+    if 0 in shifts:
+        # Finite difference formula includes a term with zero shift.
+
+        if f0 is None:
+            # Ensure that the unshifted tape is appended
+            # to the gradient tapes, if not already.
+            gradient_tapes.append(tape)
+
+        # Store the unshifted coefficient. We know that
+        # it will always be the first coefficient due to processing.
+        c0 = coeffs[0]
+        shifts = shifts[1:]
+        coeffs = coeffs[1:]
+
+    method_map = choose_grad_methods(diff_methods, argnum)
+
+    for i, _ in enumerate(tape.trainable_params):
+        if i not in method_map or method_map[i] == "0":
+            # parameter has zero gradient
+            shapes.append(0)
+            continue
+
+        g_tapes = generate_shifted_tapes(tape, i, shifts * h)
+        gradient_tapes.extend(g_tapes)
+        shapes.append(len(g_tapes))
+
+    def processing_fn(results):
+        grads = []
+        start = 1 if c0 is not None and f0 is None else 0
+        r0 = f0 or results[0]
+
+        output_dims = []
+        for m in tape.measurements:
+            if m.return_type is qml.measurements.Probability:
+                output_dims.append(2 ** len(m.wires))
+            else:
+                output_dims.append(1)
+
+        for s in shapes:
+
+            if s == 0:
+                # parameter has zero gradient
+                if not isinstance(results[0], tuple):
+                    g = qml.math.zeros_like(results[0])
+                else:
+                    g = []
+                    for i in output_dims:
+                        g.append(qml.math.zeros(i))
+                grads.append(g)
+                continue
+
+            res = results[start : start + s]
+            start = start + s
+            # compute the linear combination of results and coefficients
+
+            # First compute the multiplication with coeff
+            linear_comb = []
+            for i, c in enumerate(coeffs):
+                if isinstance(res[i], (tuple, list)):
+                    elem = [r * c for r in res[i]]
+                else:
+                    elem = [res[i] * c]
+                linear_comb.append(elem)
+
+            # Second add all the terms for each measurement separately
+            pre_grads = []
+
+            for i in range(len(tape.measurements)):
+                elem = sum(r[i] for r in linear_comb)
+                pre_grads.append(elem)

```

Comment:
Any reason why we're not using `qml.math.tensordot` here like in the parameter shift transform? Seems like it would remove a lot of this code.

Reply:
The thing is that if you use directly tensordot on the results it will not work as you will create ragged tensors. So you have to separate the results, but my real problem is that I am unable to differentiate the transform with Torch and TF, I just get 0 all the time.

```
pre_grads = []

if len(tape.measurements) == 1:
    res = qml.math.array(res, like=qml.math.get_interface(res[0]))
    r_i = qml.math.tensordot(coeffs, res, [[0], [0]])
    pre_grads.append(r_i)
else:
    for i in range(len(tape.measurements)):
        r_ = qml.math.array([r[i] for r in res], like=qml.math.get_interface(res[0]))
        r_i = qml.math.tensordot(coeffs, r_, [[0], [0]])
        pre_grads.append(r_i)
```